### DISCUSSION TITLE
Soliciting opinions on the deprecation of TLS 1.0/1.1 in OpenSSL

### DISCUSSION THREAD

**Discussion by nhorman** [openssl/openssl#25427] [General Discussion]
Recently, OpenSSL proposed the deprecation of TLS 1.0/1.1 and solicited community feedback on the idea on the oss-security mailing list.

Feedback on the proposal was generally split down the middle, with half of the respondents indicating immediate depreciation with near-term removal was acceptable, while the remainder of the respondents with affirmative opinions noted that they represent, or know of products whose environment disallowed updating to TLS1.2 or later, and would need to re-enable the deprecated features for the foreseeable future.

Given this feedback, OpenSSL would like to propose deprecation of these protocol versions (allowing for build time re-enablement) in OpenSSL 3.5. This build time configuration option would disable the building of code supporting TLS1.0/1.1 entirely, preventing runtime enablement.  Enabling the build time variable would compile the code in, but leave it runtime disabled unless the security level was reduced to 0 (as with previous releases).  A complete removal of the TLS 1.0/1.1 code would follow in a future major release.

OpenSSL is looking for further feedback on that proposed action here.

**Comment by FdaSilvaYY** [comment]
There should be an in-between security-level 0 and 1 to just enable TLS 1.0/1.1 without getting all other relics available at 0-level.

I made to myself the same reflexion about a previous disable/downgrade (sha1 or md5 ?)

**Reply by t8m** [reply]
SHA1 and MD5 is necessary for TLS-1.0 and 1.1 (except for RSA key exchange which is problematic on its own). As the security level of SHA1 and MD5 is clearly 0 and that is not going to change, I do not think it would be even possible to revert the change putting TLS-1.0/1.1 at security level 0. Besides adding fractional security level is impossible due to API constraints.

**Reply by t8m** [reply]
BTW in practice all the relics at 0 level are inside the legacy provider, so if you do not load it or exclude it via property query when the SSL_CTX is created, you will be able to avoid them.

**Reply by FdaSilvaYY** [reply]
Ok, I agree.
Thanks for the sum-up.

**Comment by BugOfBugs** [comment]
Again you deliberately avoid mention that discussion already was 3 times:
https://github.com/openssl/openssl/issues/15335
https://github.com/openssl/project/issues/750
Only there you link, and only to "_feedback_" on mailing list.
I noted that on that list you avoid link to here.
Barely anyone on list "_represents_" anything that "_disallows_" deprecation.
You ignore my notes, and behave the same.

You do not provide reason for proposal. You do not mention whether it is in the 3 discussions. You simply abandon them without conclusion.
- Why deprecate?
- Whether?
- When?
- Why remove?..

**My proposal**:
Do not deprecate (_disable_).
If you do, ask distributors to **compile 2 options**: _disabled_, and _enabled_.
Remove only if **no** longer **maintain**, and maintenance is **due**. If no need to maintain, keep forever.
You imply that will remove in version 4. But you do **not justify** why.

If you do **not deprecate**, then **no changes** necessary.
Just extra **advice** to not use old.
Later **simply** remove, if at all.

**You propose** additional, **complex** work: to deprecate.
\- Not necessary.

@FdaSilvaYY That is the [problem with _OPENSSL_TLS_SECURITY_LEVEL_](https://github.com/openssl/openssl/issues/15335#issuecomment-2162045707) (_@SECLEVEL_). It is **implicit**. You **barely realize** what it includes.
For same reason your statement is nonsense. You **do not know** what you ask for.
Would there be **separate level** for each thing, then level would **mean nothing**.

Default _OPENSSL_TLS_SECURITY_LEVEL 2_, previously _1_, causes confusion.
[Misleading reports of errors.](https://github.com/openssl/openssl/issues/15335#issuecomment-2174003753)
New _OpenSSL_ **mysteriously fails** with old configuration.
[<code>s_<b>client</b>()</code>, and <code>s_<b>server</b>()</code> treat _SHA1_ **differently**: accept, and forbid respectively.](https://github.com/openssl/openssl/issues/15335#issuecomment-1680149982) Thus, _**TLS1**_, and _**1.1**_ do not work on **server** at **_Level 1_**. - **Contradicts documentation**. Even if they make intermediate level for you, it may not work. Because of the **bugs**. Or by design, which is likely to be one more bug.
_Level_ may be **different** between programs even on same computer. Yet [`openssl()` falsely implies as if it pertains to _Level_ at which _OpenSSL_ runs with your programs](https://github.com/openssl/openssl/issues/21643#issuecomment-1664927017).
[Should **abolish**. Or at least **default** to **0**.](https://github.com/openssl/project/issues/750#issuecomment-2246719029)

**Comment by BugOfBugs** [comment]
>@t8m security level of SHA1 and MD5 is 0

This does <b>not relate</b> to next sentence, and what @FdaSilvaYY asked.
> not possible to revert the change putting TLS-1.0/1.1 at security level 0

@FdaSilvaYY did <b>not</b> ask to <b>revert</b>.

<b>Which</b> change to <b>revert</b>? Deprecation now, or assignment <i>TLS1</i> to <i>OPENSSL_TLS_SECURITY_LEVEL 2</i>?

<i><b>TLS1</b></i> does work at <b><i>Level</i> 0</b>.
As I reported, fails at <i>Level 1</i>, and greater. <b>Bug.</b> Thus, you <b>must</b> assign it to <i><b>Level 0</b></i>, and <b>correct documentation</b>.
Or promote <i>SHA1</i> to <i><b>Level 2</b></i>.

The mismatch is characteristic confusing nature of levels in general.
Better enable things <b>individually</b>.
<b>Document</b> their supposed levels. Advise. So, users may choose by knowledge, rather than magic.
> fractional security level is impossible due to API constraints

<b>What</b> constraints?

You are <b>wrong</b>. It depends on <b>definition</b> of <i>fraction</i>, and <i>constraints</i>.
<a href="https://github.com/openssl/project/issues/750#issuecomment-2246719029">You yourself are master of <b>redefinition</b></a>...
>all relics at 0 level are inside the legacy provider

What do you mean by <b>all</b>? Plenty?
<code>man OSSL_PROVIDER-legacy</code>,
<code>openssl list -provider legacy -<b>digest</b>-algorithms</code>
list but <b>few</b>. 5 algorithms.
<code>man()</code> does <b>not</b> specify whether they are <b>digest</b>, or <b>cipher</b>.
<code>-<b>cipher</b>-algorithms</code> lists more. Missing from <code>man()</code>.

Not sure which providers are enabled. Perhaps both. Lists as <b>active</b>:
<code>openssl list -provider <b>legacy</b> -provider <b>default</b> -providers</code>
Though, <i>legacy</i> does not slip in easily. <i>OpenSSL</i> chooses <i>default</i> by default.

<i>Legacy</i> is the only part we know about <i>Level 0</i>. Not all.
About other <i>Levels</i> we know even less.
If you want to disable "<i>relics</i>" of any <i>Level</i>, you are not sure what they are.
Try <code>openssl list</code>. But it does not explain header "<i><b>Legacy:</b></i>". Simply shows it, only for some <i>options</i>. Maybe means <b>absent</b>.
Can not select <i>Level</i>. To be sure have to test, or examine code.

Can select <b>only</b> with sub-commands that use list of ciphers: <code>ciphers</code>, <code>s_server</code>, <code>s_client</code>:
<code>openssl ciphers -s -v "@<b>SECLEVEL</b>=0:ALL</code>
To list ciphers for both providers insert:
<code>-provider <b>legacy</b> -provider <b>default</b></code>

These are <b>bugs</b> in code and documentation.

If you <b>remove</b> some ciphers, protocols, <b>less security</b> may result, rather than more. Some clients connect <b>without encryption</b>, if server does not match. You would have to <b>prohibit</b> this for all.
